### PR TITLE
Token Manager refactor and Group app deprecation

### LINKED ISSUE #17: Multisig app
At the present moment the most widely used governance mechanism is through a multisig. 

Different alternatives for implementation:
- Use a widely used trusted multisig like Gnosis.
- Extend Group app to require `n` confirmations before performing action (currently Group app is a 1 out of `m` multisig)

I personally think it is a better idea to implement our own solution for this on top of the Group app. Another question if we go down this route is whether we have two separate apps (Group and Multisig) or the current Group functionality is just a Multisig with only one required signature.

### LINKED ISSUE #50: Remove staking logic from Token Manager (wrapped mode)
The 'wrapped mode' makes the Token Manager too complicated. The wrapping/staking logic would be better separated into other app/s. Token Manager only purpose should be to manage a token it has controller power over. 

A generic staking contract could be done that could also work for other usecases. 

Token Manager loses the notion of having two 'modes' (https://github.com/aragon/aragon-apps/blob/master/apps/token-manager/contracts/TokenManager.sol#L66), and everything specific to wrapped mode needs to be gone.

### LINKED ISSUE #51: Allow TokenManager to make tokens non-transferable
Add a flag to the TokenManager app to block all token transfers besides initial minting, assignments by the TokenManager and burning of tokens.

### LINKED ISSUE #52: Limit how many tokens can an address have
After #51, adding a limit on how many tokens can a particular holder have, the Token + Voting app can start to resemble a board (a multisig).

This limit would not apply to how many tokens the Token Manager itself is allowed to have.

### LINKED ISSUE #53: Keep track of all token holders in Token Manager
This is an extremely short term solution that will allow the Token Manager frontend to load much quicker. 

Given #51, which allows to have boards with just tokens, a way to quickly a list of all token holders is important.

Adding this will add ~45k gas every time a new holder is added. Expensive but might be worth it in UX terms in the short term.

### LINKED ISSUE #54: Deprecate Group app
#51 + #52 + #53 allow the same functionality as the current Group app, plus a multisig (closes #17) and different board structures when combined with a voting app. All this, plus removing a lot of code!

Membership to a group would be expressed by ownership of a non-transferable token. Actions as a group member can be forwarded with the forward function of the token manager!

Note: will also close #35.

### PR DESCRIPTION
**Pull request by izqui**
- [x] Removes wrapped token logic (closes #50)
- [x] Adds the ability to make non-transferable tokens (closes #51)
- [x] Adds max tokens per account flag (closes #52)
- [x] Keeps track of all holders if flag added on init (closes #53)
- [x] Deprecates Group app (closes #54)
- [x] Tests

Allows implicit multisig functionality when paired with Voting app, closes #17

Pending doc update: https://github.com/aragon/aragon-wiki/issues/103

### CODE REVIEW

**Comment by izqui** [comment]
It removes also the frontend for the Group app, lets not forget about that because a lot of it will be reusable

File: apps/token-manager/test/tokenmanager.js [hunk 1]
```diff
@@ -31,17 +33,125 @@ contract('Token Manager', accounts => {
...
+        })
+
+        it('can assign up to limit', async () => {
+            await tokenManager.issue(limit)
+            await tokenManager.assign(holder, limit)
+
+            assert.equal(await token.balanceOf(holder), limit, 'should have tokens')
+        })
+
+        it('cannot assing more than limit', async () => {
```
Comment (sohkai) [inline apps/token-manager/test/tokenmanager.js]: Actually kind of hilarious, can we `:%s/assign/assing/g`? 😲

**Comment by izqui** [comment]
> But in general, can we use braces for all conditional blocks when the statement's not on the same line, even if it's just one statement? Pretty please? 

The freaking linter forces us to do this. I like doing them in one line `if (x) doY()`. We can turn off the linter for this, what do you think?

File: apps/token-manager/contracts/TokenManager.sol [hunk 2]
```diff
@@ -200,6 +191,38 @@ contract TokenManager is App, Initializable, TokenController, EVMCallScriptRunne
...
+    * @dev Notifies the controller about a token transfer allowing the
+    *      controller to decide whether to allow it or react if desired
+    * @param _from The origin of the transfer
+    * @param _to The destination of the transfer
+    * @param _amount The amount of the transfer
+    * @return False if the controller does not authorize the transfer
+    */
+    function onTransfer(address _from, address _to, uint _amount) public constant returns (bool) {
+        bool allowInc = isBalanceIncreaseAllowed(_to, _amount);
+        bool checkVesting = transferrableBalance(_from, now) >= _amount;
```
[the remark this answers is not in the document]
Comment (izqui) [reply apps/token-manager/contracts/TokenManager.sol]: 👍 english is hard ;)

File: apps/token-manager/contracts/TokenManager.sol [hunk 3]
```diff
@@ -200,6 +191,38 @@ contract TokenManager is App, Initializable, TokenController, EVMCallScriptRunne
...
+
+        bool allowTransfer = isTokenManager || canTransfer;
+
+        if (!allowTransfer)
+            return false;
+
+        if (logHolders)
+            _logHolderIfNeeded(_to);
+
+        return true;
```
[the remark this answers is not in the document]
Comment (izqui) [reply apps/token-manager/contracts/TokenManager.sol]: Yes! This is great, this used to be a big chain of ORs so we got the advantage. Adding the full way ;)

File: apps/token-manager/contracts/TokenManager.sol [hunk 4]
```diff
@@ -288,15 +311,23 @@ contract TokenManager is App, Initializable, TokenController, EVMCallScriptRunne
...
         require(token.transfer(_receiver, _amount));
     }
 
-    function _burn(address _holder, uint256 _amount) internal {
-        token.destroyTokens(_holder, _amount);  // minime.destroyTokens() never returns false
-    }
-
     function _mint(address _receiver, uint256 _amount) internal {
         token.generateTokens(_receiver, _amount); // minime.generateTokens() never returns false
+        if (logHolders)
```
[the remark this answers is not in the document]
Comment (izqui) [reply apps/token-manager/contracts/TokenManager.sol]: yes

File: apps/token-manager/test/tokenmanager.js (hunk 1 above)
Comment (izqui) [reply apps/token-manager/test/tokenmanager.js]: ENGLISH IS HARD ;)

**Comment by izqui** [comment]
I believe all comments to be addressed now. @sohkai please merge!

**Comment by sohkai** [comment]
> I like doing them in one line `if (x) doY()`

👍 

> The freaking linter forces us to do this.

Hmm, with the newest version (solium@1.1.2, which apparently does include an `lbrace` fix), I don't get any linter errors when using with or without the brace in the conditionals.

**Pull request merged**